Enhance "from_node" node meta to track source recursively

yushangdi opened the pull request:
Summary:
Change the "from_node" node meta format to be able to track the provenance of nodes recursively.

The new "from_node" format is a a list node NodeSource:

```
class NodeSource:
	self.node_name: str
	self.target: str
	self.graph_id: int
	self.pass_name: str
	self.action: str
	self.from_node: List[NoedSource]
```

This is in preparation for the inductor provenance tracking. For background, the inductor provenance tracking doc: https://docs.google.com/document/d/1dGh9myqNhywmbfP0Quzx_f04bghDFlj8cawj8MopiO8/edit?fbclid=IwZXh0bgNhZW0CMTEAAR0jUQ0Tf4ROLDED8Y_eIzrU0KVZVdRmyIQLp-avt-kGRPI_VgYVNyjH_q0_aem_HCQ_pxHDiwOkO9mQyWB2-g&tab=t.0 (internal only),

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export -- -r test_unflatten_multiple_graphs_state
buck run mode/dev-nosan caffe2/test:fx -- -r node_source
```

Differential Revision: D66737916




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

avikchaudhuri commented on torch/fx/traceback.py:
```diff
@@ -15,12 +20,76 @@
+    target: str
+    graph_id: int
+    pass_name: str
+    action: str
```
Is it possible to have an enum here for action?

yushangdi replied:
Fixed now!

avikchaudhuri commented on torch/fx/traceback.py:
```diff
@@ -15,12 +20,76 @@
+    If node `a` is created from node `b`, then `a.meta["from_node"]` may contain NodeSource(b).
+    """
+
+    node_name: str
```
why not just `name`?

yushangdi replied:
> why not just `name`?

I could also just make it name. Just there's also a "pass_name". So I thought it might be good to specify that there are two types of names explicitly, node_name and pass_name.

Also `node_name` might not be available. In inductor (will be added in follow-up Diffs), sometimes we don't have a specific node source, but only pass_name. For example, in some graph pass (not an Interpreter pass, but a pre_grad or post_grad inductor graph pass), we made some changes. We might not know what is the specific source node, but at least we know which pass added this node. So `node_name` might be empty. 


Do you think just `name` is better?

avikchaudhuri replied:
Yeah. Also if the node specific info is missing I wonder if it would be better to add another level of indirection instead of filling in defaults like `-1`, `""`, etc .

avikchaudhuri commented on torch/fx/traceback.py:
```diff
@@ -15,12 +20,76 @@
+                if "from_node" in node.meta
+                else []
+            )
+        else:
```
What's the point of letting `node` be optional?

yushangdi replied:
See comment above. Sometimes we don't know the source node, we just know a source pass.

avikchaudhuri commented on torch/fx/traceback.py:
```diff
@@ -15,12 +20,76 @@
+        return self.print_readable()
+
+    def print_readable(self, indent=0):
+        if indent > 36:
```
Feels a bit arbitrary? In any case I'd probably keep `indent` = 0, 1, ... and multiply by 4 when creating the string so that you can guard on the level instead of 4 x level.

yushangdi replied:
Fixed now!

avikchaudhuri reviewed (approved):
Nits

yushangdi committed:
Enhance "from_node" node meta to track source recursively (#142066)

Summary:

Change the "from_node" node meta format to be able to track the provenance of nodes recursively.

The new "from_node" format is a a list node NodeSource. The old "from_node" is a list of tuples `(node.name, node.target)`

```
class NodeSource:
    class NodeInfo:
        def __init__(self, name: str, target: str, graph_id: int):
            self.name = name
            self.target = target
            self.graph_id = graph_id

    pass_name: str
    action: Optional["NodeSourceAction"]
    from_node: List["NodeSource"]
    node_info: Optional["NodeInfo"]
```


We also add a "pass_name" field to indicate which pass created the node from the source. For now the "pass_name" is only used in fx.Interpreter.

These changes will be used for the inductor provenance tracking.

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export -- -r test_unflatten_multiple_graphs_state
buck run mode/dev-nosan  fbcode//caffe2/test:fx -- -r node_source
buck run mode/dev-nosan  fbcode//caffe2/test:fx -- -r graph_provenance
buck run mode/dev-nosan  fbcode//caffe2/test/quantization:test_quantization -- -r test_simple_metadata_porting
```

Reviewed By: avikchaudhuri

Differential Revision: D66737916

yushangdi commented:
@pytorchbot merge -i
